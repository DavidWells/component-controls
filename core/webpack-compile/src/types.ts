import * as webpack from 'webpack';
import { RuleTypes } from '@component-controls/webpack-configs';
import { LoadingStore } from '@component-controls/loader';

/**
 * configuration properties for compile and run
 */
export interface CompileProps {
  /**
   * webpack configuration object
   */
  webPack?: webpack.Configuration;
  /**
   * a list of webpack configuration presets from webpack-configs packages
   */
  presets?: RuleTypes;
  /**
   * path to the configuration file e.g : '.storybook'
   */
  configPath?: string;

  /** public output folder for the bundle */
  outputFolder?: string;
}

/**
 * return type from compile and watch functions
 */
export interface CompileResults {
  /**
   * the stories store
   */
  store: LoadingStore;
  /**
   * the webpack stats object
   */
  stats: webpack.Stats;
}

/**
 * adds webpack WatchOptions to the Compiler options
 */
export type WatchProps = {
  watchOptions?: webpack.ICompiler.WatchOptions;
} & CompileProps;
