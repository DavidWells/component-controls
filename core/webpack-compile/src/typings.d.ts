declare module '@component-controls/loader/plugin';
