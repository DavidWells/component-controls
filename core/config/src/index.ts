import * as path from 'path';
import * as fs from 'fs';
import { sync as globSync } from 'glob';
import yargs from 'yargs';
import { Configuration } from '@component-controls/specification';

export const configFileNames = ['main.js', 'main.json', 'main.ts'];

export interface ConfigrationResult {
  config: Configuration;
  configPath: string;
}
/**
 * return the configration object
 * command line accepts -c/ -config parameter for config path
 * the config file is assumed named main.js/main.ts
 * @param baseFolder project folder to start the searh with
 */
export const getConfiguration = (
  baseFolder: string,
  args: string[] = process.argv,
): ConfigrationResult | undefined => {
  const argv = yargs(args)
    .option('config', {
      alias: 'c',
      description: 'configuration folder',
      type: 'string',
    })
    .help()
    .alias('help', 'h').argv;

  const configPath = argv.config
    ? path.resolve(baseFolder, argv.config)
    : baseFolder;
  const configFiles = fs.readdirSync(configPath);
  const configFile = configFiles.find(file =>
    configFileNames.includes(file.toLowerCase()),
  );

  return configFile
    ? {
        config: require(path.resolve(configPath, configFile)),
        configPath,
      }
    : undefined;
};

/**
 * find the story files out of a configuration file
 * using glob for the regex file search
 * @param param0 configuration info return by getConfiguration
 */
export const extractStories = ({
  config,
  configPath,
}: ConfigrationResult): string[] | undefined => {
  const stories =
    config && config.stories
      ? config.stories.reduce((acc: string[], storyRg: string) => {
          const regex = storyRg.replace(/\((.*?)\)/g, '+($1)');
          return [...acc, ...globSync(path.resolve(configPath, regex))];
        }, [])
      : undefined;
  return stories;
};
