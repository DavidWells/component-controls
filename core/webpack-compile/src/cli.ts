import { compile, watch } from './index';
import { cliArgs, CliArgTypes, ArgOptions } from './args';
import { LogLevel } from '@component-controls/logger';

export { cliArgs, defaultCliArgs, ArgOptions } from './args';

export const getArgs = (options?: ArgOptions): CliArgTypes => {
  const argv = cliArgs(options);
  return argv.help().alias('help', 'h').argv;
};
export const run = async (options?: ArgOptions) => {
  const argv = getArgs(options);
  process.env.NODE_ENV = argv.mode;
  const exec = argv.watch ? watch : compile;
  await exec({
    configPath: argv.config,
    presets: argv.presets,
    bundleName: argv.bundle,
    distFolder: argv.dist,
    staticFolder: argv.static,
    logOptions: {
      logLevel: argv.loglevel as LogLevel,
    },
  });
};
