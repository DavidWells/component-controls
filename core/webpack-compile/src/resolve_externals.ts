import { Configuration } from 'webpack';

export interface ResolveExternalsConfig {
  resolve: { alias: { [key: string]: string } };
  externals: Configuration['externals'];
}

export interface ExternalProps {
  lib: string;
  externalName?: string | { [key: string]: string };
}

export class ResolveExternals {
  private _config: ResolveExternalsConfig;
  constructor(config: ResolveExternalsConfig) {
    this._config = config;
  }
  addAlias = (props: ExternalProps): void => {
    const { lib, externalName } = props;
    if (typeof this._config.externals !== 'object') {
      this._config.externals = {};
    }
    (this._config.externals as Record<string, unknown>)[lib] =
      externalName || lib;
  };
}
