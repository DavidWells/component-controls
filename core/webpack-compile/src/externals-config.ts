import { ExternalProps } from './resolve_externals';
export const defaultExternals: ExternalProps[] = [
  {
    lib: 'react',
    externalName: {
      commonjs: 'react',
      commonjs2: 'react',
      amd: 'React',
      root: 'React',
    },
  },
  {
    lib: 'react-dom',
    externalName: {
      commonjs: 'react-dom',
      commonjs2: 'react-dom',
      amd: 'ReactDOM',
      root: 'ReactDOM',
    },
  },
  {
    lib: 'prop-types',
    externalName: {
      commonjs: 'prop-types',
      commonjs2: 'prop-types',
      amd: 'PropTypes',
      root: 'PropTypes',
    },
  },
  { lib: 'regenerator-runtime' },
  {
    lib: 'lodash',
    externalName: {
      commonjs: 'lodash',
      commonjs2: 'lodash',
      amd: 'lodash',
      root: '_',
    },
  },

  { lib: 'theme-ui' },
  { lib: '@theme-ui/presets' },
  { lib: '@theme-ui/core' },
  { lib: '@theme-ui/css' },

  { lib: '@component-controls/core' },
  { lib: '@component-controls/store' },
  { lib: '@component-controls/blocks' },
  { lib: '@component-controls/components' },

  { lib: 'broadcast-channel' },

  { lib: 'react-helmet' },
  { lib: 'react-table' },
  { lib: 'react-colorful' },
  { lib: 'react-google-charts' },
  { lib: 'react-tabs' },
  { lib: 'react-switch' },
  { lib: 'react-popper' },
  { lib: 'react-popper-tooltip' },
  { lib: 'react-animate-height' },
  { lib: 'polished' },
  { lib: 'tinycolor2' },

  { lib: 'prism-react-renderer/themes' },
  { lib: 'prism-react-renderer' },

  { lib: '@primer/octicons-react' },
  { lib: '@mdx-js/react' },

  { lib: 'markdown-to-jsx' },

  { lib: 'axe-core' },
  { lib: 'faker/locale/en_US' },
  { lib: 'fuse.js' },
];
