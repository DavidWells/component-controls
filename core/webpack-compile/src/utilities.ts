import webpack, {
  Configuration,
  Compiler,
  Stats,
  HotModuleReplacementPlugin,
} from 'webpack';
import * as path from 'path';
const chalk = require('chalk');
import LoaderPlugin from '@component-controls/loader/plugin';
import {
  mergeWebpackConfig,
  deepMergeWebpackConfig,
} from '@component-controls/webpack-configs';
import { loadConfiguration } from '@component-controls/config';
import { CompileProps, CompileResults } from './types';
import { ResolveExternals, ResolveExternalsConfig } from './resolve_externals';
import { defaultExternals } from './externals-config';

export type CompileRunProps = CompileProps & {
  /**
   * webpack mode
   */
  mode: Configuration['mode'];
};

const createConfig = (options: CompileRunProps): webpack.Configuration => {
  const { webPack, presets, configPath, mode, outputFolder } = options;
  const distFolder = path.resolve(__dirname, '../dist');
  const plugins = [
    new LoaderPlugin({
      config: configPath,
      escapeOutput: mode === 'development',
    }),
    mode === 'development' && new HotModuleReplacementPlugin({}),
  ].filter(Boolean);
  const webpackConfig = mergeWebpackConfig(
    {
      mode,
      entry: {
        stories: require.resolve(
          '@component-controls/loader/story-store-data.js',
        ),
      },
      output: {
        path: distFolder,
        filename: 'bundle.js',
        libraryTarget: 'umd',
        globalObject: 'this',
      },
      resolve: {
        alias: {},
      },
      externals: {},
      plugins,
      ...(webPack || {}),
    },
    presets,
    { outputFolder, distFolder },
  );

  //add all the aliases to avoid double loading of packages
  const alias = new ResolveExternals(webpackConfig as ResolveExternalsConfig);
  defaultExternals.forEach(a => alias.addAlias(a));

  const runConfig = loadConfiguration(process.cwd(), configPath);

  const userWebpackConfig =
    runConfig?.config &&
    (runConfig.config.webpack || runConfig.config.finalWebpack);

  if (!userWebpackConfig) {
    return webpackConfig;
  }
  return typeof userWebpackConfig === 'function'
    ? userWebpackConfig(webpackConfig, options)
    : deepMergeWebpackConfig(webpackConfig, userWebpackConfig);
};

export const runCompiler = (
  run: (
    compiler: Compiler,
    callback: (err: Error, stats: Stats) => void,
  ) => void,

  props: CompileRunProps,
): Promise<CompileResults> => {
  return new Promise((resolve, reject) => {
    //@ts-ignore
    const compiler = webpack(createConfig(props));
    run(compiler, (err, stats) => {
      if (err) {
        console.error(err);
        return reject(err);
      }
      if (stats.hasErrors() || stats.hasWarnings()) {
        const error = stats.toString({
          errorDetails: true,
          warnings: true,
        });
        console.error(error);
        return reject(error);
      }

      const { store } = require('@component-controls/loader/store');
      console.log(
        chalk.bgRgb(244, 147, 66)(`compiled ${store.stores.length} documents`),
        require.resolve('@component-controls/loader/store'),
      );
      resolve({ store, stats });
    });
  });
};
