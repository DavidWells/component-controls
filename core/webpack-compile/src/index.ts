const chalk = require('chalk');
import { CompileProps, CompileResults, WatchProps } from './types';
import { runCompiler } from './utilities';

export { getBundleName, defBundleName } from './utilities';
export * from './types';
/**
 * compile the stories with webpack
 * returns the stories store object
 */
export const compile = ({
  webPack,
  presets,
  configPath,
  staticFolder,
  distFolder,
  bundleName,
}: CompileProps): Promise<CompileResults> => {
  console.log(
    chalk.bgRgb(244, 147, 66)('@start compilation'),
    'optimized build',
  );
  return runCompiler((compiler, callback) => compiler.run(callback), {
    webPack,
    mode: 'production',
    presets,
    configPath,
    staticFolder,
    distFolder,
    bundleName,
  });
};

/**
 * compile the stories with webpack and launch watching for changes
 * returns the stories store object
 */
export const watch = ({
  webPack,
  presets,
  configPath,
  watchOptions,
  staticFolder,
  distFolder,
  bundleName,
}: WatchProps): Promise<CompileResults> => {
  console.log(
    chalk.bgRgb(244, 147, 66)('@start compilation'),
    'development mode watch',
  );
  return runCompiler(
    (compiler, callback) => compiler.watch({ ...watchOptions }, callback),
    {
      webPack,
      mode: 'development',
      presets,
      configPath,
      staticFolder,
      distFolder,
      bundleName,
    },
  );
};
