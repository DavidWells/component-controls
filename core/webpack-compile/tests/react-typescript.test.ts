import * as path from 'path';
import { compile } from '../src/index';

expect.addSnapshotSerializer({
  test: val => typeof val === 'object' && val instanceof String,
  print: val => {
    return `"${val.toString()}"`;
  },
});

describe('test typescript react settings', () => {
  it('compile', async () => {
    const { store } = await compile({
      presets: ['react', 'react-docgen-typescript'],
      configPath: path.resolve(__dirname, 'fixtures'),
    });
    expect(store).toMatchSnapshot();
  }, 30000);
});
