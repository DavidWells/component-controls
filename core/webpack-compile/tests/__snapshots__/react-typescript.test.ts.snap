// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test typescript react settings compile 1`] = `
Object {
  "buildConfig": Object {
    "categories": Array [
      "author",
      "tags",
    ],
    "ignore": Array [
      "readme.md",
      "changelog.md",
      "code_of_conduct.md",
      "contributing.md",
      "license.md",
    ],
    "pages": Object {
      "author": Object {
        "basePath": "authors/",
      },
      "blog": Object {
        "basePath": "blogs/",
      },
      "page": Object {
        "basePath": "pages/",
      },
      "story": Object {
        "basePath": "docs/",
        "sideNav": Object {
          "collapseSingle": true,
          "storyPaths": true,
        },
        "tabs": Array [
          Object {
            "route": "page",
          },
        ],
      },
      "tags": Object {
        "basePath": "tags/",
      },
    },
    "stories": Array [
      "../../../../ui/components/src/ActionBar/*.stories.(js|jsx|tsx|mdx)",
    ],
  },
  "components": Object {
    "fded7dbaa855ea6869a794f877eab680": Object {
      "from": "./ActionBar",
      "importedName": "ActionBar",
      "imports": Object {
        "react": Array [
          Object {
            "importedName": "FC",
            "name": "FC",
          },
          Object {
            "importedName": "useMemo",
            "name": "useMemo",
          },
        ],
        "theme-ui": Array [
          Object {
            "importedName": "Box",
            "name": "Box",
          },
          Object {
            "importedName": "Button",
            "name": "Button",
          },
          Object {
            "importedName": "jsx",
            "name": "jsx",
          },
        ],
      },
      "info": Object {
        "description": "a strip of actions to be attached to a container
the action items contain the labels and click event handler
actions can accept an order prop, and can also be superimposed",
        "displayName": "ActionBar",
        "methods": Array [],
        "props": Object {
          "actions": Object {
            "defaultValue": "[]",
            "description": "collection of action items",
            "parentName": "ActionBarProps",
            "type": Object {
              "name": "ActionItems",
              "raw": "ActionItems",
            },
          },
          "themeKey": Object {
            "defaultValue": "actionbar",
            "description": "custom layouts from the theme, defaults to 'actionbar'",
            "parentName": "ActionBarProps",
            "type": Object {
              "name": "string",
              "raw": "string",
            },
          },
        },
      },
      "loc": Object {
        "end": Object {
          "column": 1,
          "line": 75,
        },
        "start": Object {
          "column": 45,
          "line": 25,
        },
      },
      "name": "ActionBar",
      "package": "fded7dbaa855ea6869a794f877eab680",
      "request": "/Users/atanasster/component-controls/ui/components/src/ActionBar/ActionBar.tsx",
      "source": "/** @jsx jsx */
import { FC, useMemo } from 'react';
import { Box, Button, jsx } from 'theme-ui';
import { getSortedActions, ActionItems } from './utils';
import { Link } from '../Link';

export interface ActionBarProps {
  /**
   * collection of action items
   */
  actions?: ActionItems;

  /**
   * custom layouts from the theme, defaults to 'actionbar'
   */
  themeKey?: string;
}

/**
 * a strip of actions to be attached to a container
 * the action items contain the labels and click event handler
 * actions can accept an order prop, and can also be superimposed
 *
 */
export const ActionBar: FC<ActionBarProps> = ({
  actions = [],
  themeKey = 'actionbar',
}) => {
  const items = useMemo(() => {
    const sortedItems = getSortedActions(actions);
    return sortedItems.map(
      ({ node, onClick, 'aria-label': ariaLabel, group, href }, index) => {
        const nextGroup =
          index < sortedItems.length - 1 ? sortedItems[index + 1].group : group;
        return (
          <Box
            key={\`\${typeof node === 'string' ? node : 'item'}_\${index}\`}
            variant={\`\${themeKey}.item\`}
            sx={{
              mr: index === 0 ? 1 : 0,
              ml: nextGroup !== group || group === undefined ? 2 : 1,
            }}
          >
            {typeof node === 'string' ? (
              href ? (
                <Link
                  variant={\`\${themeKey}.link\`}
                  href={href}
                  aria-label={ariaLabel}
                >
                  {node}
                </Link>
              ) : (
                <Button
                  variant={\`\${themeKey}.button\`}
                  onClick={onClick}
                  aria-label={ariaLabel}
                >
                  {node}
                </Button>
              )
            ) : (
              node
            )}
          </Box>
        );
      },
    );
  }, [actions, themeKey]);
  return (
    <Box variant={\`\${themeKey}.container\`}>
      <Box variant={\`\${themeKey}.inner\`}>{items}</Box>
    </Box>
  );
};
",
    },
  },
  "config": undefined,
  "packages": Object {
    "8c5a495b6c81e803332610dad9861480": Object {
      "dependencies": Object {
        "@primer/octicons-react": "^10.0.0",
        "@theme-ui/css": "^0.4.0-rc.1",
        "@theme-ui/match-media": "^0.4.0-rc.1",
        "@theme-ui/presets": "^0.4.0-rc.1",
        "copy-to-clipboard": "^3.2.1",
        "fast-memoize": "^2.5.2",
        "markdown-to-jsx": "^6.11.0",
        "mdx-utils": "*",
        "prism-react-renderer": "^1.0.2",
        "react": "^16.13.1",
        "react-animate-height": "^2.0.20",
        "react-dom": "^16.13.1",
        "react-popper-tooltip": "^3.1.0",
        "react-switch": "^5.0.1",
        "react-table": "^7.0.0",
        "react-tabs": "^3.1.0",
        "scroll-into-view-if-needed": "^2.2.25",
        "theme-ui": "^0.4.0-rc.1",
      },
      "devDependencies": Object {
        "@component-controls/jest-snapshots": "^1.27.0",
        "@component-controls/ts-markdown-docs": "^1.21.0",
        "@types/markdown-to-jsx": "^6.11.0",
        "@types/react-table": "^7.0.10",
        "@types/react-tabs": "^2.3.1",
        "cross-env": "^5.2.1",
        "eslint": "^6.5.1",
      },
      "fileHash": "8c5a495b6c81e803332610dad9861480",
      "name": "@component-controls/components",
      "peerDependencies": Object {
        "@theme-ui/css": ">= 0.4.0-rc.1",
        "react": ">= 16.8.0",
        "react-dom": ">= 16.8.0",
        "theme-ui": ">= 0.4.0-rc.1",
      },
      "repository": Object {
        "browse": "https://github.com/ccontrols/component-controls/tree/master/ui/components/src/ActionBar/ActionBar.stories.tsx",
        "docs": "https://github.com/ccontrols/component-controls/tree/master#readme",
        "issues": "https://github.com/ccontrols/component-controls/issues",
      },
      "version": "1.27.0",
    },
    "fded7dbaa855ea6869a794f877eab680": Object {
      "dependencies": Object {
        "@primer/octicons-react": "^10.0.0",
        "@theme-ui/css": "^0.4.0-rc.1",
        "@theme-ui/match-media": "^0.4.0-rc.1",
        "@theme-ui/presets": "^0.4.0-rc.1",
        "copy-to-clipboard": "^3.2.1",
        "fast-memoize": "^2.5.2",
        "markdown-to-jsx": "^6.11.0",
        "mdx-utils": "*",
        "prism-react-renderer": "^1.0.2",
        "react": "^16.13.1",
        "react-animate-height": "^2.0.20",
        "react-dom": "^16.13.1",
        "react-popper-tooltip": "^3.1.0",
        "react-switch": "^5.0.1",
        "react-table": "^7.0.0",
        "react-tabs": "^3.1.0",
        "scroll-into-view-if-needed": "^2.2.25",
        "theme-ui": "^0.4.0-rc.1",
      },
      "devDependencies": Object {
        "@component-controls/jest-snapshots": "^1.27.0",
        "@component-controls/ts-markdown-docs": "^1.21.0",
        "@types/markdown-to-jsx": "^6.11.0",
        "@types/react-table": "^7.0.10",
        "@types/react-tabs": "^2.3.1",
        "cross-env": "^5.2.1",
        "eslint": "^6.5.1",
      },
      "fileHash": "fded7dbaa855ea6869a794f877eab680",
      "name": "@component-controls/components",
      "peerDependencies": Object {
        "@theme-ui/css": ">= 0.4.0-rc.1",
        "react": ">= 16.8.0",
        "react-dom": ">= 16.8.0",
        "theme-ui": ">= 0.4.0-rc.1",
      },
      "repository": Object {
        "browse": "https://github.com/ccontrols/component-controls/tree/master/ui/components/src/ActionBar/ActionBar.tsx",
        "docs": "https://github.com/ccontrols/component-controls/tree/master#readme",
        "issues": "https://github.com/ccontrols/component-controls/issues",
      },
      "version": "1.27.0",
    },
  },
  "stores": Array [
    Object {
      "doc": Object {
        "component": "ActionBar",
        "components": Object {},
        "componentsLookup": Object {
          "ActionBar": "fded7dbaa855ea6869a794f877eab680",
        },
        "date": "2020-08-18T06:58:47.265Z",
        "dateModified": "2020-08-18T06:58:47.271Z",
        "fileName": "/Users/atanasster/component-controls/ui/components/src/ActionBar/ActionBar.stories.tsx",
        "package": "8c5a495b6c81e803332610dad9861480",
        "title": "Components/ActionBar",
      },
      "filePath": "/Users/atanasster/component-controls/ui/components/src/ActionBar/ActionBar.stories.tsx",
      "stories": Object {
        "groupEnd": Object {
          "arguments": Array [],
          "id": "groupEnd",
          "loc": Object {
            "end": Object {
              "column": 1,
              "line": 123,
            },
            "start": Object {
              "column": 24,
              "line": 105,
            },
          },
          "name": "groupEnd",
          "renderFn": [Function],
          "source": "() => (
  <Container>
    <ActionBar
      actions={[
        {
          node: 'item 1',
        },
        {
          node: 'item 2',
          group: '1',
        },
        {
          node: 'item 3',
          group: '1',
        },
      ]}
    />
  </Container>
)",
        },
        "groupStart": Object {
          "arguments": Array [],
          "id": "groupStart",
          "loc": Object {
            "end": Object {
              "column": 1,
              "line": 143,
            },
            "start": Object {
              "column": 26,
              "line": 125,
            },
          },
          "name": "groupStart",
          "renderFn": [Function],
          "source": "() => (
  <Container>
    <ActionBar
      actions={[
        {
          node: 'item 1',
          group: '1',
        },
        {
          node: 'item 2',
          group: '1',
        },
        {
          node: 'item 3',
        },
      ]}
    />
  </Container>
)",
        },
        "link": Object {
          "arguments": Array [],
          "id": "link",
          "loc": Object {
            "end": Object {
              "column": 1,
              "line": 62,
            },
            "start": Object {
              "column": 20,
              "line": 52,
            },
          },
          "name": "link",
          "renderFn": [Function],
          "source": "() => (
  <Container>
    <ActionBar
      actions={[
        {
          node: <ExternalLink href=\\"https://google.com\\">google</ExternalLink>,
        },
      ]}
    />
  </Container>
)",
        },
        "order": Object {
          "arguments": Array [],
          "id": "order",
          "loc": Object {
            "end": Object {
              "column": 1,
              "line": 81,
            },
            "start": Object {
              "column": 21,
              "line": 64,
            },
          },
          "name": "order",
          "renderFn": [Function],
          "source": "() => (
  <Container>
    <ActionBar
      actions={[
        {
          node: 'action 1',
          onClick: () => console.log('clicked'),
          order: 1,
        },
        {
          node: <ExternalLink href=\\"https://google.com\\">google</ExternalLink>,
          onClick: () => console.log('clicked'),
          order: 0,
        },
      ]}
    />
  </Container>
)",
        },
        "override": Object {
          "arguments": Array [],
          "id": "override",
          "loc": Object {
            "end": Object {
              "column": 1,
              "line": 103,
            },
            "start": Object {
              "column": 24,
              "line": 83,
            },
          },
          "name": "override",
          "renderFn": [Function],
          "source": "() => (
  <Container>
    <ActionBar
      actions={[
        {
          node: 'action 1',
          onClick: () => console.log('clicked'),
          id: 'copy',
        },
        {
          node: <ExternalLink href=\\"https://google.com\\">google</ExternalLink>,
          onClick: () => console.log('clicked'),
        },
        {
          node: 'Copy',
          id: 'copy',
        },
      ]}
    />
  </Container>
)",
        },
        "overview": Object {
          "arguments": Array [
            Object {
              "loc": Object {
                "end": Object {
                  "column": 29,
                  "line": 0,
                },
                "start": Object {
                  "column": 1,
                  "line": 0,
                },
              },
              "value": Array [
                Object {
                  "loc": Object {
                    "end": Object {
                      "column": 11,
                      "line": 0,
                    },
                    "start": Object {
                      "column": 3,
                      "line": 0,
                    },
                  },
                  "name": "themeKey",
                  "usage": Array [
                    Object {
                      "loc": Object {
                        "end": Object {
                          "column": 24,
                          "line": 3,
                        },
                        "start": Object {
                          "column": 16,
                          "line": 3,
                        },
                      },
                    },
                  ],
                  "value": "themeKey",
                },
              ],
            },
          ],
          "controls": Object {
            "themeKey": Object {
              "options": Array [
                "actionbar",
                "toolbar",
                "footer",
              ],
              "type": "options",
              "value": "actionbar",
            },
          },
          "id": "overview",
          "loc": Object {
            "end": Object {
              "column": 1,
              "line": 40,
            },
            "start": Object {
              "column": 24,
              "line": 24,
            },
          },
          "name": "overview",
          "renderFn": [Function],
          "source": "({ themeKey }: ActionBarProps) => (
  <Container>
    <ActionBar
      themeKey={themeKey}
      actions={[
        {
          node: 'action 1',
          onClick: () => console.log('clicked'),
        },
        {
          node: <ExternalLink href=\\"https://google.com\\">google</ExternalLink>,
          onClick: () => console.log('clicked'),
        },
      ]}
    />
  </Container>
)",
        },
      },
    },
  ],
}
`;
